pytorch_lightning/callbacks/stochastic_weight_avg.py

```diff
@@ -194,14 +225,14 @@ def on_train_epoch_start(self, trainer: "pl.Trainer", pl_module: "pl.LightningMo
             else:
                 trainer.lr_scheduler_configs.append(default_scheduler_cfg)
 
-            self.n_averaged = torch.tensor(0, dtype=torch.long, device=pl_module.device)
+            if self.n_averaged is None:
+                self.n_averaged = torch.tensor(self._init_n_averaged, dtype=torch.long, device=pl_module.device)
 
         if self.swa_start <= trainer.current_epoch <= self.swa_end:
             self.update_parameters(self._average_model, pl_module, self.n_averaged, self._avg_fn)
```

Comment:
this should check whether we already updated parameters for the current epoch as this hook can get called again for the same epoch when the trainer restarts.

Reply:
How would that happen? Is that when restoring from a mid-epoch checkpoint?

Reply:
I believe doing that will fix `test_swa_resume_training_from_checkpoint[4]`

Reply:
That happens because when you save a checkpoint `on_train_epoch_end` (ModelCheckpoint default), the epoch has technically not finished. See description in https://github.com/PyTorchLightning/pytorch-lightning/pull/8578

So after it is reloaded, the FitLoop (loop that manages epochs) does one "dummy" iteration to get to the correct state. During that dummy iteration the `TrainingEpochLoop` is skipped as you observed above.

This has the side effect that epoch-level hooks get called again on restoration. So `update_parameters` would be getting called again for the previous epoch.

Reply:
Ok thanks for the explanation, that makes sense. I'm not sure that matches up with what I was observing though. Eg. if I crash during epoch 3, then the checkpoint from epoch 2 is loaded and training restarts, but trainer.current_epoch for the first dummy iteration that doesn't do any training is 3, whereas if I'm understanding you correctly, it should be 2?

Reply:
I've made the change you suggested here but as suspected, it doesn't actually fix the tests. When resuming from a checkpoint, the first "dummy" epoch number is the epoch after the checkpoint was saved, not the epoch from the checkpoint. So things look like this in the case where we crash on epoch number 3:
```
epoch 0: 5 batches
epoch 1: 5 batches
epoch 2: 5 batches
epoch 3: crash
epoch 3: 0 batches
epoch 4: 5 batches
epoch 5: 5 batches (extra epoch due to batchnorm)
```

So epoch 3 effectively never happens. This is because when loading the checkpoint data, `fit_loop.epoch_progress.current` in the checkpoint is `{'ready': 3, 'completed': 2, 'started': 3, 'processed': 3}`, so the setter for `FitLoop.restarting` overrides `completed` and sets it to 3. If I remove [that line](https://github.com/PyTorchLightning/pytorch-lightning/pull/8578/files#diff-a3f9154e7c08e7e6f2c703a4a483679193a2f4268cf1cf495d466a872e7cd5fdR145), then all my tests pass, but there are some other checkpoint related tests that start to fail (test_checkpoint_repeated_strategy_extended and test_trainer_model_hook_system_fit_no_val_and_resume).

The logic in that `restarting` setter seems to contradict the idea that there should be a dummy epoch with no batches though.

Reply:
Exploring the fix in https://github.com/PyTorchLightning/pytorch-lightning/pull/12429